File: backend/app/schemas/explainability.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class FeatureImportanceItem(BaseModel):
    feature: str
    importance: float
    rank: int
    direction: str | None = None
    original_feature: str | None = None


class FeatureContribution(FeatureImportanceItem):
    contribution: float


class LocalExplanation(BaseModel):
    row_index: int
    entity_id: str | None = None
    prediction: Any
    prediction_probability: float | None = None
    predicted_value: float | str | None = None
    probabilities_by_class: dict[str, float] = {}
    baseline_value: float | None = None
    method: str
    feature_contributions: list[FeatureContribution]
    top_positive_factors: list[FeatureContribution]
    top_negative_factors: list[FeatureContribution]
    warnings: list[str] = []


class ExplainabilityResult(BaseModel):
    dataset_id: str
    model_id: str
    problem_type: str
    explanation_method: str
    model_artifact: str
    generated_at: str
    global_importance: list[FeatureImportanceItem]
    local_explanations: list[LocalExplanation]
    warnings: list[str] = []

File: frontend/src/types/explainability.ts
export interface FeatureImportanceItem {
  feature: string;
  importance: number;
  rank: number;
  direction: "positive" | "negative" | "neutral" | null;
  original_feature: string | null;
}

export interface FeatureContribution extends FeatureImportanceItem {
  contribution: number;
}

export interface LocalExplanation {
  row_index: number;
  entity_id: string | null;
  prediction: string | number | boolean | null;
  prediction_probability: number | null;
  predicted_value: number | string | null;
  probabilities_by_class: Record<string, number>;
  baseline_value: number | null;
  method: string;
  feature_contributions: FeatureContribution[];
  top_positive_factors: FeatureContribution[];
  top_negative_factors: FeatureContribution[];
  warnings: string[];
}

export interface ExplainabilityResult {
  dataset_id: string;
  model_id: string;
  problem_type: "classification" | "regression" | "clustering";
  explanation_method: string;
  model_artifact: string;
  generated_at: string;
  global_importance: FeatureImportanceItem[];
  local_explanations: LocalExplanation[];
  warnings: string[];
}
